File: BerkeleyAlgorithm.java
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

public class BerkeleyAlgorithm {
    private static final long MILLIS_PER_SECOND = 1000;

    public static void berkeleyAlgo(String servertime, String time1, String time2) {
        System.out.println("Server Clock = " + servertime);
        System.out.println("Client Clock = " + time1);
        System.out.println("Client clock 2 = " + time2);

        SimpleDateFormat sdf = new SimpleDateFormat("mm:ss");
        try {
            long s = sdf.parse(servertime).getTime();
            long t1 = sdf.parse(time1).getTime();
            long t2 = sdf.parse(time2).getTime();

            long st1 = t1 - s;
            System.out.println("t1 - s  = " + st1 / 1000);

            long st2 = t2 - s;
            System.out.println("t2 - s  = " + st2 / 1000);

            long aveg = ((st1 + st2) / 1000);
            System.out.println("(st1 + st2) / 2 = " + aveg / 2);

            long adjserver = (aveg / 2) * 1000 + s;
            long adj_t1 = (aveg / 2) * 1000 - st1;
            long adj_t2 = (aveg / 2) * 1000 - st2;

            System.out.println("t1 adjustment = " + adj_t1 / 1000);
            System.out.println("t2 adjustment = " + adj_t2 / 1000);

            System.out.println("Synchronized Server Clock = " + sdf.format(new Date(adjserver)));
            System.out.println("Synchronized Client 1 Clock = " + sdf.format(new Date(t1 + adj_t1)));
            System.out.println("Synchronized Client 2 Clock = " + sdf.format(new Date(t2 + adj_t2)));

        } catch (ParseException e) {
            System.err.println("Error parsing time: " + e.getMessage());
        }
    }

    public static void main(String[] args) {
        berkeleyAlgo("10:00", "10:05", "10:03");
    }
}

File: BerkeleyAlgorithm.py
from datetime import datetime, timedelta

def berkeley_algo(server_time, time1, time2):
    print("Server Clock =", server_time)
    print("Client Clock 1=", time1)
    print("Client Clock 2 =", time2)
    
    time_format = "%H:%M"
    try:
        s = datetime.strptime(server_time, time_format)   
        t1 = datetime.strptime(time1, time_format)       
        t2 = datetime.strptime(time2, time_format)
        
        st1 = (t1 - s).total_seconds()
        st2 = (t2 - s).total_seconds()
        print("t1 - s =", st1)
        print("t2 - s =", st2)
        
        avg_offset = (st1 + st2) / 2
        print("(st1 + st2) / 2 =", avg_offset)

        adj_server = s + timedelta(seconds=avg_offset)
        adj_t1 = t1 - timedelta(seconds=st1 - avg_offset)
        adj_t2 = t2 - timedelta(seconds=st2 - avg_offset)
        
        print("Synchronized Server Clock =", adj_server.strftime(time_format))
        print("Synchronized Client 1 Clock =", adj_t1.strftime(time_format))
        print("Synchronized Client 2 Clock =", adj_t2.strftime(time_format))
        
    except ValueError as e:
        print("Error parsing time:", e)

berkeley_algo("10:00", "10:05", "10:03")
        